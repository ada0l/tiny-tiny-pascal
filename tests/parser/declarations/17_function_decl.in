function max(num1, num2: integer): integer;
var result: integer;
begin
   if (num1 > num2) then result := num1
   else result := num2;
   max := result;
end;

function pi_calc(const total: Integer): Real;
var inside: Integer = 0; 
    i: Integer;
    point: record x, y: Real end;
    x, y: Real;
begin
    for i := 0 to total do begin
        point.x := random();
        point.y := random();
        if sqrt(point.x * point.x + point.y * point.y) < 1 then inside += 1;
    end;
    pi_calc := inside / total * 4 ;
end;

begin end.
