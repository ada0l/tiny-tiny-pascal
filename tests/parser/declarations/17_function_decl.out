main
   program name:
      without name
   decls:
      function decl
         max 
            param
               without modifier
               num1 
               num2 
               type:
                  integer 
            type:
               integer 
         decls:
            vars:
               var
                  result 
                  type:
                     integer 
                  without exp
         block:
            compounds statements
               if
                  >
                     num1 
                     num2 
                  :=
                     result 
                     num1 
                  else:
                     :=
                     result 
                     num2 
               :=
                  max 
                  result 
      function decl
         pi_calc 
            param
               const 
               total 
               type:
                  Integer 
            type:
               Real 
         decls:
            vars:
               var
                  inside 
                  type:
                     Integer 
                  0
               var
                  i 
                  type:
                     Integer 
                  without exp
               var
                  point 
                  record
                     fields
                        type:
                           Real 
                        x 
                        y 
                  without exp
               var
                  x 
                  y 
                  type:
                     Real 
                  without exp
         block:
            compounds statements
               for
                  i 
                  0
                  to 
                  total 
                  compounds statements
                     :=
                        point 
                           x 
                        random 
                     :=
                        point 
                           y 
                        random 
                     if
                        <
                           sqrt 
                              +
                                 *
                                    point 
                                       x 
                                    point 
                                       x 
                                 *
                                    point 
                                       y 
                                    point 
                                       y 
                           1
                        +=
                           inside 
                           1
                        else:
                           empty
               :=
                  pi_calc 
                  *
                     /
                        inside 
                        total 
                     4
   block:
      compounds statements